### PR TITLE
Fully Implementing ProjectionCalculator

### PR DESCRIPTION
**Pull request by mlarson02**
This PR completes the implementation of the ProjectionCalculator by taking code from the TrackletCalculatorBase.cc and firmware-hls ProjectionCalculator.h to calculate TrackletProjections from TrackletParameter input. Also makes the required changes in the Tracklet class and TrackletConfigBuilder to remove functionality in TrackletProcessor to calculate projections (now is only calculating parameters). 

Some code to calculate projections has been left in the TrackletCalculatorBase.cc to allow the TrackletCalculatorDisplaced.cc to continue working.

### CODE REVIEW

File: L1Trigger/TrackFindingTracklet/interface/ProjectionCalculator.h [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include "L1Trigger/TrackFindingTracklet/interface/TrackletLUT.h"
 #include "L1Trigger/TrackFindingTracklet/interface/TrackletProjectionsMemory.h"
 #include "L1Trigger/TrackFindingTracklet/interface/TrackletParametersMemory.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletCalculatorBase.h"
+#include "L1Trigger/TrackFindingTracklet/interface/TrackletConfigBuilder.h"
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/interface/ProjectionCalculator.h]: Does these need to be in the header file?

File: L1Trigger/TrackFindingTracklet/src/ProjectionCalculator.cc [hunk 2]
```diff
@@ -61,32 +142,287 @@ void ProjectionCalculator::addInput(MemoryBase* memory, string input) {
...
+    
+
+    // L6
+    ir = settings_.irmean(5);
+    projLayer(ir, irinv, iphi0, it, iz0, izr_LD[5], iphi_LD[5]);
+    valid_zmin = izr_LD[5] >= zmin;
+    valid_zmax = izr_LD[5] < zmax;
+    valid_phimax = iphi_LD[5] < phimax;
+    valid_phimin = iphi_LD[5] > phimin;
+    valid_LD[5] = valid_zmin & valid_zmax & valid_phimax & valid_phimin;
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/src/ProjectionCalculator.cc]: Could this be written in a more compact form with a loop?

**Review by aryd — COMMENTED** [review]
A few questions/suggestions added

File: L1Trigger/TrackFindingTracklet/interface/ProjectionCalculator.h (hunk 1 above)
Comment (mlarson02) [reply L1Trigger/TrackFindingTracklet/interface/ProjectionCalculator.h]: TrackletCalculatorBase.h is not necessary and will be removed, but TrackletConfigBuilder.h is to use the iMergedTCStr function to get the page the projections need to be written to (this seemed the best way to do this).

**Comment by tomalin** [comment]
@aryd As Mark Larson has now stopped working on this (but is still contactable), could you please add a comment here explaining what prevents this being merged?

**Comment by aryd** [comment]
Ian,

 I think this should also be ready to merge. There is a little bit of an issue with a slight reduction in the tracking efficiency that Mark did not understand. I have not had a chance to look at this. But I think we should merge this and then sort out the inefficiency before this gets used in an official release.

 -Anders


Anders Ryd
***@***.******@***.***>



On Jul 30, 2024, at 6:28 PM, Ian Tomalin ***@***.******@***.***>> wrote:



@aryd<https://github.com/aryd> As Mark Larson has now stopped working on this (but is still contactable), could you please add a comment here explaining what prevents this being merged?

**Comment by tomalin** [comment]
By the way, this fails git CI because the git CI assumes that the PR is being made to cms-L1TK:L1TK-dev-14_0_0_pre2, so rebases the branch to this branch.

**Pull request merged**